### PR TITLE
Rhum v2.0

### LINKED ISSUE #75: Rhum v2.0 (Release Date: TBD)
## Tasks

- [x] CLI options (#74)
- [x] Script to generate test file (#44)
- [x] Mock not taking arguments (#28)
- [x] Ability to assert what arguments a method was called with (#21)
- [x] Skip functionality (#16)
- [x] real-time results (#78)
- [x] `make` subcommand (#44)
- [x] ~Ability to test non-public members (#19)~ (Won't do. Users should use the Reflect API.)
- [x] ~`--ignore` option (#82)~ (Logic is no longer needed. `run` command checks files for Rhum namespace now.)
- [ ] Rhum lifecycle diagram (#81)
- [ ] Documentation (https://github.com/drashland/website/issues/227)
    - [ ] Add v2.0 docs
    - [ ] Add migration guide
    - [ ] Archive v1.0 docs, but let it be accessible through an Archive section

### PR DESCRIPTION
**Pull request by crookse**
Closes #75 

### Usage

Clone the repo.

```
git clone https://github.com/drashland/rhum.git
git checkout rhum-cli
```

Install Rhum.

```
deno install --allow-read --allow-run -f ./cli.ts
```

Display help menu.

```
rhum [SUBCOMMAND] [OPTIONS|DENO FLAGS] [dir/test file]
```

### Features
- Improved local and CI output
- Supports deno flags (e.g., `--allow-write`, `--allow-net`)
- Directly states where the assertion error occurred (e.g., test file name, line, and column)
    ```
    tests/integration/asserts_test.ts
        Deno std asserts
            PASS asserts
            FAIL assertEquals
    
    AssertionError: Values are not equal:
    
        [Diff] Actual / Expected
    
    -   false
    +   true
    
    The above assertion error occurred in:
    
       asserts_test.ts:9:20
    
    Test Results: 8 passed; 1 failed; 0 skipped
    ```
- Displays what test file is being run before any test suites and test cases run
- Displays errors that occur within the framework (should they occur)
- Adds method: `mock.methods.{someMethod}.wasLastCalledWith([arg1, arg2, etc.])`
    ```typescript
    const mockMathService = Rhum
      .mock(MathService)
      .create();
    mockMathService.add(1, 1);
    Rhum.asserts.assert(
      mockMathService.methods.add.wasLastCalledWith([1, 1]),
    );
    mockMathService.add(2, 1);
    Rhum.asserts.assert(
      mockMathService.methods.add.wasLastCalledWith([2, 1]),
    );
    ```
- Adds method: `mock.methods.{someMethod}.wasCalledTimes(numberOfTimes)`
    ```typescript
    const mockMathService = Rhum
      .mock(MathService)
      .create();
    const mockTestObject = Rhum
      .mock(TestObject)
      .withConstructorArgs("has mocked math service", mockMathService)
      .create();
    Rhum.asserts.assert(mockMathService.methods.add.wasCalledTimes(0));
    mockTestObject.sum(1, 1); // calls mockMathService.add(1, 1)
    Rhum.asserts.assert(mockMathService.methods.add.wasCalledTimes(1));
    ```
- Adds method: `mock.methods.{someMethod}.lastReturned(returnValue)`
    ```typescript
    const mockMathService = Rhum
      .mock(MathService)
      .create();
    mockMathService.add(1, 1);
    Rhum.asserts.assert(
      mockMathService.methods.add.lastReturned(2),
    );
    mockMathService.add(2, 1);
    Rhum.asserts.assert(
      mockMathService.methods.add.lastReturned(3),
    );
    ```

### Substitutions
- Uses `LoggerService` instead of `logInfo()`, `logError()`, etc.

### Removes
-  `ServerRequest` mock
- `mock.calls`

### Chores
- Uses one bumper script to handle bumping for pre-release and when deno updates (see `console/bumper_ci_service.ts`)
- Tests the bumper script to test that regex patterns are correct
- Removes duplicate regex patterns in bumper script

### CI output

Screen Shot 2020-10-17 at 09 36 04

### Local output (with error and actual/expected diff)

Screen Shot 2020-10-22 at 09 23 07

### CODE REVIEW

**Comment by ebebbington** [comment]
@crookse not sure if you were aware but the windows tests essentially timed out, something hanged and GH's vm gave up after 360 minutes

**Comment by crookse** [comment]
LOL 360 minutes. dang. ok. gonna have to look into that.

File: src/cli/commands/make.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import { LoggerService } from "../../../deps.ts";
+
+const encoder = new TextEncoder();
+
+const testTemplate: Uint8Array = encoder.encode(
+  `import { Rhum } from "https://deno.land/x/rhum@v1.1.4/mod.ts";
```
Comment (ebebbington) [inline src/cli/commands/make.ts:6]: use `version` from `mod.ts` here? `import ... @${version}`

**Comment by crookse** [comment]
opening a rhum-v2 branch PR

**Comment by crookse** [comment]
see #110

**Pull request closed without merging**